Pure SSH-based protocol

bk2204 opened the pull request:
This is an implementation of a pure SSH-based protocol.  It uses the proposal outlined in the `docs/proposals` directory, with a few changes, which are added to the documentation.

The implementation here is complete.  It should be fully functional.

This implementation is being tested against [Scutiger's `git-lfs-transfer` binary](https://github.com/bk2204/scutiger), which is our reference implementation.  That binary will be used as part of our CI jobs in the future to validate that we continue to work correctly with it.  It is built from the latest crate version now that it is published on crates.io.

The goal is to have the code fall back to `git-lfs-authenticate` and use the HTTP-based protocol with SSH authentication if the `git-lfs-transfer` binary fails for whatever reason.

Fixes #1044

chrisd8088 commented on t/t-locks.sh:
```diff
@@ -76,6 +76,34 @@ begin_test "list a single lock (SSH)"
+  set -e
+
+  reponame="locks-list-ssh-pure"
+  setup_pure_ssh
```
The `"batch transfers with ssh endpoint (git-lfs-transfer)"` test in `t/t-batch-transfer.sh` has this `setup_pure_ssh` line immediately after the initial `set -e` and separated with whitespace, which I liked because it helps visually distinguish this test from the one above it, so maybe it's worth doing the same here too?

chrisd8088 commented on t/testhelpers.sh:
```diff
@@ -222,6 +248,33 @@ refute_server_lock() {
 
+# refute that a lock with the given ID exists on the test server
+refute_server_lock_ssh() {
+  local reponame="$1"
+  local id="$2"
+  local refspec="$3"
+  local destination="$(canonical_path "$REMOTEDIR/$reponame.git")"
+
+  (
+    pktize_text 'version 1'
+    pktize_flush
+    pktize_text 'list-lock'
+    pktize_text "id=$id"
+    pktize_text "refname=$refname"
+    pktize_flush
+    pktize_text 'quit'
+    pktize_flush
+  ) | lfs-ssh-echo git@127.0.0.1 "git-lfs-transfer '$destination' download" 2>&1 |
+    tee output.log
+
+  grep "status 200" output.log
```
Most of this function seems like it's identical to most of `assert_server_lock_ssh()`, so I wonder if common portions could be refactored into a shared sub-function?

bk2204 replied:
Sure.

chrisd8088 commented on t/t-locks.sh:
```diff
@@ -76,6 +76,34 @@ begin_test "list a single lock (SSH)"
+  grep "f.dat" locks.log
+  grep "lfs-ssh-echo.*git-lfs-transfer .*$reponame.git download" trace.log
+
+  git lfs unlock --id "$id"
+  refute_server_lock_ssh "$reponame" "$id" "refs/heads/main"
```
The other tests in `t/t-locks.sh` don't seem to exercise the unlock functionality, which makes this new test a little special in that it's testing unlocking over SSH too, as well as it's stated purpose of `"list a single lock (SSH; git-lfs-transfer)"`.

Should tests for the the basic lock and unlock functionality over SSH be implemented in `t/t-lock.sh` and `t/t-unlock.sh`, respectively, leaving this new test in `t/t-locks.sh` just for checking the `git lfs locks` listing functionality?  (Obviously the test _also_ implicitly has to set up a lock first, which means `git lfs lock` will be exercised too, but that's more of a side-effect of setting up the test conditions than a properly separate test of the lock-taking functionality.)

bk2204 replied:
Yeah, I think we can do that.

chrisd8088 commented on tq/ssh.go:
```diff
@@ -196,8 +198,7 @@ func (a *SSHAdapter) download(t *Transfer, cb ProgressCallback) error {
 // download starts a download. dlFile is expected to be an existing file open in RW mode
-func (a *SSHAdapter) downloadOne(t *Transfer, cb ProgressCallback, dlFile *os.File) error {
-	conn := a.transfer.Connection(0)
+func (a *SSHAdapter) downloadOne(t *Transfer, conn *ssh.PktlineConnection, cb ProgressCallback, dlFile *os.File) error {
```
What about changing this function signature to `doDownload(t *Transfer, conn *ssh.PktlineConnection, f *os.File, cb ProgressCallback)` to mirror the `doUpload()` function?  (The comment should become `f is expected ...` if this change was made.)  I feel like that might be nice because they serve such similar purposes, are both internal to this source file, and because I like having the callback parameter at the end of the signature, I guess because it seems less important to me than the file handle.

bk2204 replied:
We can certainly do that.

chrisd8088 commented on tq/ssh.go:
```diff
@@ -185,7 +187,7 @@ func (a *SSHAdapter) download(t *Transfer, cb ProgressCallback) error {
 	}()
 
-	err = a.downloadOne(t, cb, f)
+	err = a.downloadOne(t, conn, cb, f)
```
Does the `defer` function cover the error case as well?  Maybe this could just be `return a.downloadOne(t, conn, cb, f)`?

bk2204 replied:
Yes, I think that's possible.

chrisd8088 reviewed (changes requested):
This is an amazing contribution, thank you so much!  It will be greatly appreciated by a significant portion of the Git LFS community, and it will be a major new feature that distinguishes the 3.0 release.  🚀

I hope my comments and handful of suggestions are clear; please let me know if any are not!

Thank you again for working on this so thoroughly and diligently over many months.  ❤️

bk2204 commented on git/filter_process_scanner.go:
```diff
@@ -172,14 +173,14 @@ func (o *FilterProcessScanner) Err() error { return o.err }
 	req := &Request{
 		Header:  make(map[string]string),
-		Payload: &pktlineReader{pl: o.pl},
+		Payload: pktline.NewPktlineReaderFromPktline(o.pl, 32768),
```
Yes, that's correct.  It probably makes sense to increase this to 64k, since I'm pretty sure that Git will send larger packets.  Scutiger's `git-lfs-transfer` will not (it prefers 32k chunks because they're the largest power of 2 supported in a pkt-line), but Git almost certainly will.

bk2204 commented on ssh/pktline.go:
```diff
@@ -0,0 +1,86 @@
+
+func pktlineReader(p Pktline) io.Reader {
+	if pl, ok := p.(*pktline.Pktline); ok {
+		return pktline.NewPktlineReaderFromPktline(pl, 65536)
+	}
+	tp := p.(*TraceablePktline)
+	return pktline.NewPktlineReaderFromPktline(tp.pl, 65536)
```
Nope.  I'll adjust them.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,275 @@
+	return conn.pl.WriteFlush()
+}
+
+func (conn *PktlineConnection) SendMessageWithArguments(command string, args []string, lines []string) error {
```
Sounds good.  I can make that change.

bk2204 commented on lfsapi/lfsapi.go:
```diff
@@ -51,8 +51,8 @@ func (c *Client) Context() lfshttp.Context {
 func (c *Client) SSHTransfer(operation, remote string) *ssh.SSHTransfer {
-	endpoint := c.Endpoints.RemoteEndpoint(operation, remote)
-	if len(endpoint.SSHMetadata.UserAndHost) > 0 {
+	endpoint := c.Endpoints.Endpoint(operation, remote)
+	if len(operation) > 0 && len(endpoint.SSHMetadata.UserAndHost) > 0 {
```
I'll make that change.  The reason we want to use `Endpoint` over `RemoteEndpoint` is because we want to use `lfs.url`.  I'll update the original patch with a suitable commit message.

bk2204 commented on locking/api.go:
```diff
@@ -313,3 +313,53 @@ func NewUser(name string) *User {
+		c.useTransfer = false
+		lclient = &httpLockClient{Client: c.client}
+	}
+	c.lclient = &lclient
```
You're completely correct here.  The actual situation we have that I've seen is two requests: a `Search()` and then a `Lock()`.  Inconveniently, that's a download first and then an upload.  I haven't seen three here, although I'm happy to implement caching for this case.  If nothing else, that will simplify the logic here, which I think is going to be desirable.

bk2204 commented on locking/api_test.go:
```diff
@@ -63,7 +63,7 @@ func TestAPILock(t *testing.T) {
 	lc := &lockClient{Client: c}
 	lockRes, res, err := lc.Lock("", &lockRequest{Path: "request", Ref: &lockRef{Name: "refs/heads/master"}})
 	require.Nil(t, err)
-	assert.Equal(t, 200, res.StatusCode)
+	assert.Equal(t, 200, res)
```
Sure.

bk2204 commented on locking/ssh.go:
```diff
@@ -0,0 +1,219 @@
+					return lock, "", fmt.Errorf("lock response: invalid locked-at: %s", entry)
+				}
+			}
+		}
```
The JSON parsing doesn't guarantee the fields are present; it uses the zero values if the fields are absent.  However, I do agree some sort of checking here seems useful, so I've added some in a fixup patch.

bk2204 commented on locking/ssh.go:
```diff
@@ -0,0 +1,219 @@
+func (c *sshLockClient) parseListLockResponse(status int, args []string, text []string) (all []Lock, ours []Lock, theirs []Lock, nextCursor string, message string, err error) {
+	type lockData struct {
+		lock Lock
+		id   string
```
We can definitely remove it.

bk2204 commented on locking/ssh.go:
```diff
@@ -0,0 +1,219 @@
+		}
+		for _, entry := range text {
+			values := strings.SplitN(entry, " ", 3)
+			if len(values) < 2 || (values[0] != "lock" && len(values) < 3) {
+				return nil, nil, nil, "", "", fmt.Errorf("lock response: invalid response: %q", entry)
+			}
+			if values[0] != "lock" && (last == nil || last.id != values[1]) {
+				return nil, nil, nil, "", "", fmt.Errorf("lock response: interspersed response: %q", entry)
+			}
+			if values[0] == "lock" {
+				locks[values[1]] = &lockData{id: values[1]}
+				last = locks[values[1]]
+				last.lock.Id = values[1]
+			} else if values[0] == "path" {
+				last.lock.Path = values[2]
+			} else if values[0] == "ownername" {
+				last.lock.Owner = &User{}
+				last.lock.Owner.Name = values[2]
+			} else if values[0] == "owner" {
+				last.who = owner(values[2])
+			} else if values[0] == "locked-at" {
+				last.lock.LockedAt, err = time.Parse(time.RFC3339, values[2])
+				if err != nil {
+					return nil, nil, nil, "", "", fmt.Errorf("lock response: invalid locked-at: %s", entry)
+				}
+			}
```
This looks like a nice improvement.  I've adopted it with a few minor changes and a credit and I'll test it.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,275 @@
+	for {
+		s, pktLen, err := conn.pl.ReadPacketTextWithLength()
+		if err != nil {
+			return 0, nil, nil
```
Yup, it should.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,275 @@
+				ok := false
+				if strings.HasPrefix(s, "status ") {
+					status, err = strconv.Atoi(s[7:])
+					ok = err != nil
+				}
+				if ok {
```
Yes, thanks for catching that.  The variable is bizarrely named for the logic.  Will fix.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,275 @@
+		case seenDelim:
+			args = append(args, s)
+		default:
+			if strings.HasPrefix(s, "status ") {
+				status, err = strconv.Atoi(s[7:])
+				if err != nil {
+					return 0, nil, errors.NewProtocolError(fmt.Sprintf("expected status line, got %q", s), err)
+				}
+				if !delim {
+					seenDelim = true
+				}
+			}
+			if !seenStatus {
+				seenStatus = true
+				ok := false
+				if strings.HasPrefix(s, "status ") {
+					status, err = strconv.Atoi(s[7:])
+					ok = err != nil
+				}
+				if ok {
+					return 0, nil, errors.NewProtocolError(fmt.Sprintf("expected status line, got %q", s), err)
+				}
+			}
```
Good point.  I think this was a refactoring error.

bk2204 commented on ssh/ssh.go:
```diff
@@ -47,14 +56,59 @@ func parseShellCommand(command string, existing string) (ssh string, cmd string,
+}
+
+func getVariant(osEnv config.Environment, gitEnv config.Environment, basessh string) sshVariant {
+	variant, ok := osEnv.Get("GIT_SSH_VARIANT")
+	autodetect, val := findVariant(variant)
+	if ok {
+		if autodetect {
+			return autodetectVariant(osEnv, gitEnv, basessh)
+		}
+		return val
+	}
+	variant, ok = gitEnv.Get("ssh.variant")
+	if autodetect, val := findVariant(variant); ok {
+		if autodetect {
+			return autodetectVariant(osEnv, gitEnv, basessh)
+		}
+		return val
+	}
+	return autodetectVariant(osEnv, gitEnv, basessh)
```
There was some tricky corner case I was trying to handle here, but I think I encapsulated that in the tests, so I think your change is a strict improvement here since it's much simpler.

bk2204 commented on ssh/ssh.go:
```diff
@@ -104,9 +107,36 @@ func getVariant(osEnv config.Environment, gitEnv config.Environment, basessh str
+
+func getControlDir(osEnv config.Environment) (string, error) {
+	dir := findRuntimeDir(osEnv)
+	if dir != "" {
+		dir = filepath.Join(dir, "git-lfs")
+		err := os.Mkdir(dir, 0700)
+		if err != nil {
+			// Ideally we would use errors.Is here to check against
+			// os.ErrExist, but that's not available on Go 1.11.
+			perr, ok := err.(*os.PathError)
+			if !ok || perr.Err != syscall.EEXIST {
+				return ioutil.TempDir("", "sock-*")
+			}
+		}
+		return dir, nil
+	}
+	return ioutil.TempDir("", "sock-*")
```
Good point.  Fixed.

bk2204 commented on t/cmd/lfs-ssh-echo.go:
```diff
@@ -17,35 +20,78 @@ type sshResponse struct {
+
 func main() {
 	// expect args:
 	//   lfs-ssh-echo -p PORT -- git@127.0.0.1 git-lfs-authenticate REPO OPERATION
```
Absolutely.

bk2204 commented on t/cmd/lfs-ssh-echo.go:
```diff
@@ -17,35 +20,78 @@ type sshResponse struct {
-	authLine := strings.Split(os.Args[5], " ")
-	if len(authLine) < 13 {
+	authLine := strings.Split(os.Args[offset+1], " ")
+	if len(authLine) < 2 {
 		fmt.Fprintf(os.Stderr, "bad git-lfs-authenticate line: %s\nargs: %v", authLine, os.Args)
```
I believe this needs to be `< 2` because if we invoke Git, there isn't any operation.  We get `git-receive-pack REPO` or `git-upload-pack REPO`.  Of course, if I had bothered to update the comment (which I have done since you suggested that), this would have been clearer, so my apologies for the oversight.

I think an `os.Exit(1)` is appropriate.  Having that before would have demonstrated that this was broken, but of course we didn't.  I'll explain this in the commit message as well.

chrisd8088 replied:
OK, now I think I understand what was confusing me before (along with the weird `< 13` check).  It's that the command and arguments to be executed remotely are expected to be wrapped in quotes, e.g., `lfs-ssh-echo git@127.0.0.1 "git-lfs-transfer REPO OPERATION"` or `lfs-ssh-echo -p PORT -- git@127.0.0.1 "git-lfs-authenticate REPO OPERATION"`.

That's because of this [line](https://github.com/git-lfs/git-lfs/blob/502801373173e28268b4f8417707117d4ee5fecd/lfshttp/ssh.go#L129) in what is presently `lfshttp.sshGetLFSExeAndArgs()` and this PR refactors to `ssh.GetLFSExeAndArgs()`, which packs all of the command and arguments to be executed remotely on the SSH server into a single argument to be passed to the local SSH client.

So, knowing that, I might suggest that the comments here and above still need a bit more revision, and the error message could also benefit from being made less specific to `git-lfs-authenticate`.  For the expanded comments above, just adding the quotes would help:
```go
	//   lfs-ssh-echo -p PORT -- git@127.0.0.1 "git-lfs-authenticate REPO OPERATION"
	//   lfs-ssh-echo -p PORT -- git@127.0.0.1 "git-lfs-transfer REPO OPERATION"
	//   lfs-ssh-echo -- git@127.0.0.1 "git-lfs-transfer REPO OPERATION"
	//   lfs-ssh-echo git@127.0.0.1 "git-upload-pack REPO"
	//   lfs-ssh-echo git@127.0.0.1 "git-receive-pack REPO"
```

And then for this section, maybe something like the following?
```go
	// just "COMMAND REPO OPERATION" (from the final quoted argument)
	authLine := strings.Split(os.Args[offset+1], " ")
	if len(authLine) < 2 {
		fmt.Fprintf(os.Stderr, "bad remote SSH command: %s\nargs: %v", authLine, os.Args)
		os.Exit(1)
	}
```

Oh, and it couldn't hurt to rename `authLine` to a more generic name, since it now covers remote commands other than just `git-lfs-authenticate`.

bk2204 commented on t/cmd/lfs-ssh-echo.go:
```diff
@@ -17,35 +20,78 @@ type sshResponse struct {
+	if authLine[0] == "git-lfs-transfer" || authLine[0] == "git-upload-pack" || authLine[0] == "git-receive-pack" {
+		err := spawnCommand(os.Args[offset+1])
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error running command %q: %v", authLine[0], err)
```
Yes, fixed.

bk2204 commented on t/testhelpers.sh:
```diff
@@ -808,3 +808,10 @@ setup_pure_ssh() {
+ssh_remote() {
+  local reponame="$1"
+  local destination=$(urlify "$(canonical_path "$REMOTEDIR/$reponame.git")")
+  [ -z "${destination##/*}" ] || destination="/$destination"
```
Yes, I believe those are functionally equivalent.  My goal is to prepend a slash iff it lacks one.  This is required for Windows, but things should work just fine on Unix without it.  I'll add a comment.

Also, having worked on Git, I've learned that invoking processes in shell on Windows is just eternally slow because Windows lacks a native `fork` primitive.  (Running a Unix shell under either WSL 1 or 2 is much faster as a result.)  So it's been drilled into my head to use built-in shell primitives whenever possible, since they perform much better there.

bk2204 commented on t/testhelpers.sh:
```diff
@@ -789,3 +789,22 @@ urlify() {
+    fi
+  elif [ "$GIT_DEFAULT_HASH" = sha256 ]
+  then
+      echo "Using SHA-256 repositories.  Skipping..."
```
Done.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,391 @@
+	for _, line := range lines {
+		entries := strings.Split(line, " ")
+		if len(entries) < 3 {
+			return nil, fmt.Errorf("batch response: malformed response")
```
Fixed.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,391 @@
+		if len(entries) < 3 {
+			return nil, fmt.Errorf("batch response: malformed response")
+		}
+		length := len(bRes.Objects)
+		if length == 0 || bRes.Objects[length-1].Oid != entries[0] {
+			bRes.Objects = append(bRes.Objects, &Transfer{Actions: make(map[string]*Action)})
+		}
+		transfer := bRes.Objects[len(bRes.Objects)-1]
+		transfer.Oid = entries[0]
+		transfer.Size, err = strconv.ParseInt(entries[1], 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("batch response: invalid size: %s", entries[1])
+		}
```
I don't think we can do that because your proposed line 82 essentially calls `bRes.Objects[-1]` if `length` is 0, and it will also overwrite the previous item if this line starts a new stanza.  At least, I believe it does so, although if I'm misunderstanding, do say so.

chrisd8088 replied:
No, that makes sense, my mistake!

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,391 @@
+				} else if strings.HasPrefix(entry, "expires-at=") {
+					transfer.Actions[entries[2]].ExpiresAt, err = time.Parse(time.RFC3339, entry[11:])
+					if err != nil {
+						return nil, fmt.Errorf("batch response: invalid expires-in: %s", entry)
```
Fixed.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,391 @@
+
+func configureSSHAdapter(m *Manifest) {
+	m.RegisterNewAdapterFunc("ssh", Upload, func(name string, dir Direction) Adapter {
+		bu := &SSHAdapter{newAdapterBase(m.fs, name, dir, nil), nil, m.sshTransfer}
+		bu.transferImpl = bu
+		return bu
+	})
+	m.RegisterNewAdapterFunc("ssh", Download, func(name string, dir Direction) Adapter {
+		bu := &SSHAdapter{newAdapterBase(m.fs, name, dir, nil), nil, m.sshTransfer}
+		bu.transferImpl = bu
+		return bu
```
Yeah, I can change the variable name.  I intentionally didn't replicate the `panic` call because it doesn't matter: the adapter is the same and changing the direction doesn't change the adapter.  For our upload and download adapters, it does very much matter indeed, but not here.

bk2204 commented:
Okay, I've gone through all of the comments, including the ones I didn't reply to directly, and I believe I've addressed all of the concerns unless I've specifically said I didn't think it needed to be addressed.  For the moment, I've put in some squash and fixup patches to help you in a second round.  The only changes to existing patches are those to resolve conflicts from the insertion of patches earlier into the series (which was required in order to maintain bisectable commits).

Once you're happy with these changes as they stand, I'll just squash them down and then merge.

chrisd8088 reviewed (changes requested):
Whew!  Thank you for the many, many revisions and responses!  ❤️

I think I have marked everything that looked resolved as such, leaving only a couple of comments from my original review with additional notes or questions, and I had only a few extras turn up and they should be in the review now too.

I confess I did not go through the whole body of changes from top to bottom; I thought I'd wait on that and do it as part of a final pass.

Thanks again and please do let me know if anything of my comments are unclear, or just plain nonsensical or wrong.  😄

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,269 @@
+		} else if pktLen == 1 {
+			break
+		} else {
+			if !seenStatus {
+				seenStatus = true
+				ok := false
+				if strings.HasPrefix(s, "status ") {
+					status, err = strconv.Atoi(s[7:])
+					ok = err == nil
+				}
+				if !ok {
+					return 0, nil, nil, errors.NewProtocolError(fmt.Sprintf("expected status line, got %q", s), err)
+				}
+			} else {
+				args = append(args, s)
+			}
```
Sure, sounds good.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,269 @@
+			seenDelim = true
+		case seenDelim:
+			args = append(args, s)
+		default:
+			if !seenStatus {
```
Yup, with your suggestion above, I did just that.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,269 @@
+				if !ok {
+					return 0, nil, errors.NewProtocolError(fmt.Sprintf("expected status line, got %q", s), err)
+				}
+				if !delim {
+					seenDelim = true
+				}
```
I don't think that's quite right here.  If `delim` is true, then we expect a delimiter packet and we care about the data, which would be arguments in this case.  We'll process those in the `case seenDelim` block.  If `delim` is false, then we don't want to process those, and we set it here so we don't read them.

Did you maybe read `if !delim` as `if !seenDelim`?

chrisd8088 replied:
You're quite right, I misread it as `!seenDelim` -- my apologies, and never mind!

bk2204 commented:
I think I've addressed all of your suggestions; I took all but one, where I think one of us may have a misunderstanding, which could very well be me (and if so, please say so).  I've rebased onto the main branch, but left my changes as fixup commits at the end for easy reviewing.  As mentioned before, once we're happy with where things are, I'll squash them down before merging.

chrisd8088 commented on tq/ssh.go:
```diff
@@ -0,0 +1,390 @@
+}
+
+// Begin a new batch of uploads or downloads. Call this first, followed by
+// one or more Add calls. maxConcurrency controls the number of transfers
+// that may be done at once. The passed in callback will receive updates on
```
This comment is presumably a [duplicate](https://github.com/git-lfs/git-lfs/blob/38ae601d5db0da9f94ee3e822d9bd09d4c39da7b/tq/transfer.go#L237-L238) of the one in `tq/transfer.go` for the `Adapter` interface.  Unfortunately I think that comment has been out-of-sync with the actual `Begin()` function's definition since #1774 in 2016, when `maxConcurrency int` was replaced with `cfg AdapterConfig`.  So I guess it would be good to fix both comments now (or just drop this one and fix or ignore the legacy one).

Also, I suspect it may no longer be true that "Either argument may be nil if not required by the client" and in fact at least `cfg` must be defined (but I didn't check this).

bk2204 replied:
I've updated the comment.

chrisd8088 reviewed (changes requested):
Thank you again for your patience as I slowly make my way through this major update for a third time!

For this round I tried to read through from the top as a single set of changes, instead of commit-by-commit, and noticed a few things I'd missed before -- my sincere apologies.

And a number of my comments are just that, or questions to which the answer is probably to leave things as they are; please do ignore anything which doesn't seem correct or useful or germane.

Thanks again so much!  ❤️

bk2204 commented on docs/proposals/ssh_adapter.md:
````diff
@@ -95,6 +94,23 @@ to those used if the request is made over HTTP.
+                flush-pkt
+```
+
+This command exists to help detect expected termination from unexpected
````
Fixed.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return err
+	}
+	buf := make([]byte, 32768)
```
Ah, good question.  No, I think I'd prefer to leave it as 32 KiB.  The maximum packet size is 65519 and this is the largest packet that's a power of two.  Raising this to 64 KiB would not work here; we'd get an error.

bk2204 commented on git/filter_process_scanner_test.go:
```diff
@@ -126,13 +127,13 @@ func TestFilterProcessScannerReadsRequestHeadersAndPayload(t *testing.T) {
 func TestFilterProcessScannerRejectsInvalidHeaderPackets(t *testing.T) {
 	from := bytes.NewBuffer([]byte{
-		0x30, 0x30, 0x30, 0x34, // 0004 (invalid packet length)
+		0x30, 0x30, 0x30, 0x33, // 0004 (invalid packet length)
```
The comment should be `0003`.  Packet lengths include the 4-byte length field, so `0003` is not a valid packet (it would be a special one if it existed, but it doesn't).  We used to reject `0004`, but we no longer do, since they are valid packets, although sending them is unusual (since they're empty); this is just an error in not updating the comment.

bk2204 commented on locking/ssh.go:
```diff
@@ -0,0 +1,238 @@
+
+	}
+	var lock lockResponse
+	lock.Lock, lock.Message, err = c.parseLockResponse(status, args, text)
+	return &lock, status, err
```
Yes, that's correct.  I believe that's also true for the HTTP versions.

bk2204 commented on ssh/ssh.go:
```diff
@@ -0,0 +1,197 @@
+	return subprocess.FormatForShellQuotedArgs(cmd, args)
+}
+
+func GetLFSExeAndArgs(osEnv config.Environment, gitEnv config.Environment, meta *SSHMetadata, command, operation string, multiplexDesired bool) (string, []string) {
```
Hmm, I'm not sure if I like that name better.  The two "And" components make it sound like there's three things being returned, but we have only two.

chrisd8088 replied:
Fair enough, plus as I noted, some version of this name has been around for a long time.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+	return err
+}
+
+func (conn *PktlineConnection) readCapabilities() ([]string, error) {
+	return conn.pl.ReadPacketList()
+}
```
Fixed.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return err
+	}
+	if delim {
+		err = conn.pl.WriteDelim()
+		if err != nil {
+			return err
+		}
+	}
```
In this case, the data being passed is not a set of arguments, but a set of data lines.  I'll adjust the variable name to make it clearer.

chrisd8088 replied:
Hmm ... I actually think these _are_ arguments, and `args` was appropriate, not `lines`.  (All the callers in `locking/ssh.go` and `tq/ssh.go` pass `args`, and I think those are all correctly named, reading the proposal API spec.)

I think the confusion stems from the optional delimiter, which is actually not used at all, except by `End()` above in this file when it's sending the `"quit"` message.  In that one case only is `delim` set to `true`.  But I believe, from the API spec in this PR, that the `quit-request` just looks like:
```
quit-request = PKT-LINE("quit" LF) flush-pkt
```
which I believe would mean it shouldn't be passing `delim = true`.

So my proposal would be to drop the `delim` parameter entirely here, and all of the handling for a delimiter packet in this `SendMessage()` function.  That also aligns nicely with how `ReadStatus()` now doesn't look for delimiters.  Then all the callers can drop their `false` parameters, too (or `true` in the case of the `"quit"` message).

chrisd8088 replied:
I ran the test suite with these changes made locally, and all seems to pass OK.  (I believe I've got scutiger in place.  :-)

bk2204 replied:
Yeah, I've fixed that here by removing that chunk of code, as you've suggested.

chrisd8088 replied:
Sorry, I think there's one other detail here left to consider, which is to reverse the name change from `args` to `lines`, so it goes back to being `args`.  All the callers (e.g., in `locking/ssh.go`) name the second parameter `args`, and I think that's appropriate.

bk2204 commented on locking/ssh.go:
```diff
@@ -0,0 +1,238 @@
+	conn := c.connection()
+	conn.Lock()
+	defer conn.Unlock()
+	err := conn.SendMessage("lock", args, false)
```
Yeah, `git-lfs-transfer` just reads to a flush packet.  I'll just update the docs to drop the extra delimiter packet.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+	buf := make([]byte, 32768)
+	for {
+		n, err := data.Read(buf)
+		if n > 0 {
+			conn.pl.WritePacket(buf[0:n])
+		}
+		if err != nil {
+			break
+		}
```
We should be handling the error while writing, which I'll fix.  But we handle `n > 0` first because the error can be `io.EOF`, and the documentation for `Read` indicates that we should always handle a nonzero number of bytes read before handling the error.

chrisd8088 replied:
Ah, gotcha, thanks!

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return err
+	}
+	_, _, err = conn.ReadStatus(true)
```
No, because the remote side may have hung up already and data may not have gotten appropriately flushed, so this is a best-effort scenario.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+	return conn.pl.WriteFlush()
+}
+
+func (conn *PktlineConnection) ReadStatus(delim bool) (int, []string, error) {
```
I'll drop the delimiter packet from the message.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+			return status, args, nil
+		case pktLen == 1:
+			seenDelim = true
+		case seenDelim:
+			args = append(args, s)
```
This is a great suggestion.  I've taken it.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+}
+
+// ReadStatusWithArguments reads a status, arguments, and a set of text lines.
+func (conn *PktlineConnection) ReadStatusWithArguments() (int, []string, []string, error) {
```
Fixed.

chrisd8088 replied:
Thank you!  I fear the comment above the function may still need a tweak to match the new name.

bk2204 replied:
Fixed.

bk2204 commented on docs/proposals/ssh_adapter.md:
````diff
@@ -95,6 +94,23 @@ to those used if the request is made over HTTP.
+
+```
+quit-response = PKT-LINE("status 200" LF)
+                delim-pkt
````
Yup.  Already dropped.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,264 @@
+				return 0, nil, nil, errors.NewProtocolError("no status seen", nil)
+			}
+			return status, args, lines, nil
+		case pktLen == 1:
```
Fixed.

chrisd8088 replied:
This looks great in `ReadStatusWithLines()`!  I believe, IIUC, that `ReadStatusWithData()` could also benefit from the move of `else if !seenStatus { ... }` above `else if pktLen == 1 { ... }`, but I'll put that into a separate, fresh comment and mark this one as resolved.

bk2204 commented on ssh/ssh.go:
```diff
@@ -0,0 +1,197 @@
+
+// Return the executable name for ssh on this machine and the base args
+// Base args includes port settings, user/host, everything pre the command to execute
+func GetExeAndArgs(osEnv config.Environment, gitEnv config.Environment, meta *SSHMetadata, multiplexDesired bool) (exe string, baseargs []string, needShell bool) {
```
I believe that it's used in the tests, and the tests are in a different package for import loop reasons.

chrisd8088 replied:
Ah, I see the `ssh_test` package now; I forgot about that from the first review.  Never mind, then!

bk2204 commented on ssh/ssh_test.go:
```diff
@@ -0,0 +1,543 @@
+
+	exe, args := ssh.FormatArgs(ssh.GetExeAndArgs(cli.OSEnv(), cli.GitEnv(), &meta, false))
+	assert.Equal(t, plink, exe)
+	assert.Equal(t, []string{"-P", "8888", "user@foo.com"}, args)
```
Added.

chrisd8088 replied:
Looks great -- I might only suggest moving it up above the `TestSSHGetExeAndArgsTortoisePlink()` one so it immediately follows the other `TestSSHGetExeAndArgsPlinkCustomPortExplicitEnvironment*()` test functions.

bk2204 replied:
Done.

bk2204 commented on t/t-batch-transfer.sh:
```diff
@@ -123,3 +123,33 @@ begin_test "batch transfers with ssh endpoint"
+
+  sshurl=$(ssh_remote "$reponame")
+  git config lfs.url "$sshurl"
+  git lfs env
```
Removed.

bk2204 commented on t/t-batch-transfer.sh:
```diff
@@ -123,3 +123,33 @@ begin_test "batch transfers with ssh endpoint"
+  git lfs env
+
+  contents="test"
+  oid="$(calc_oid "$contents")"
```
Removed.

bk2204 commented on t/t-locks.sh:
```diff
@@ -72,10 +71,35 @@ begin_test "list a single lock (SSH)"
+  assert_server_lock_ssh "$reponame" "$id" "refs/heads/main"
+
+  GIT_TRACE=1 git lfs locks --path "f.dat" 2>trace.log | tee locks.log
+  cat trace.log
```
This is here so if the test fails it's obvious why.  We do that in several other places in the tests, so I'd like to keep it.

chrisd8088 replied:
Sure, sounds good.

bk2204 commented on tq/api.go:
```diff
@@ -32,6 +32,12 @@ type BatchResponse struct {
 	endpoint            lfshttp.Endpoint
 }
 
+type BatchTransferAdapter interface {
+	Batch(remote string, bReq *batchRequest) (*BatchResponse, error)
+	MaxRetries() int
+	SetMaxRetries(n int)
+}
+
```
Renamed.

bk2204 commented on tq/manifest.go:
```diff
@@ -109,11 +114,20 @@ func NewManifest(f *fs.Filesystem, apiClient *lfsapi.Client, operation, remote s
 
+	if sshTransfer != nil {
+		// Multiple concurrent transfers are not yet supported.
+		m.batchClientAdapter = &SSHBatchTransferAdapter{
```
Renamed.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,390 @@
+// ClearTempStorage clears any temporary files, such as unfinished downloads that
+// would otherwise be resumed
+func (a *SSHAdapter) ClearTempStorage() error {
+	return os.RemoveAll(a.tempDir())
```
Yeah, I think this cleanup happens elsewhere.  It is a little out of scope for this PR, but I'm happy to add a follow-up PR to fix it.

chrisd8088 replied:
Sounds good.  I think there are a few other incidental things I spotted while reviewing this PR and #4525 which we might lump into a followup, including some out-of-date comments elsewhere (like [this](https://github.com/git-lfs/git-lfs/blob/38ae601d5db0da9f94ee3e822d9bd09d4c39da7b/tq/transfer.go#L237-L239) and [this](https://github.com/git-lfs/git-lfs/blob/38ae601d5db0da9f94ee3e822d9bd09d4c39da7b/lfs/gitscanner_refs.go#L35-L36)) and extra [lines](https://github.com/git-lfs/git-lfs/blob/38ae601d5db0da9f94ee3e822d9bd09d4c39da7b/t/t-batch-transfer.sh#L114), all pre-existing stuff which could use a tidy-up.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,390 @@
+	// Must be dedicated to this adapter as deleted by ClearTempStorage
+	// Also make local to this repo not global, and separate to localstorage temp,
+	// which gets cleared at the end of every invocation
+	d := filepath.Join(a.fs.LFSStorageDir, "incomplete")
```
No, there isn't any reason not to share them.  The directories can be safely shared, and it's not even like both could legitimately be used in the same invocation anyway, since one will never use both adapters at the same time.  I've updated the comments.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,390 @@
+	}
+	defer f.Close()
+
+	status, _, text, err := a.doUpload(t, conn, f, cb)
```
I think I've got all of these spots now.

bk2204 commented on t/cmd/lfs-ssh-echo.go:
```diff
@@ -17,46 +20,95 @@ type sshResponse struct {
 
-	if os.Args[4] != "git@127.0.0.1" {
-		fmt.Fprintf(os.Stderr, "$4 expected \"git@127.0.0.1\", got %q", os.Args[4])
+	// just "git-lfs-authenticate REPO OPERATION" or "git-(upload|receive)-pack REPO"
```
I've updated the comment to reflect both possibilities.

chrisd8088 commented on locking/ssh.go:
```diff
@@ -0,0 +1,238 @@
+					}
+				}
+			}
+		}
```
Looks like my suggested changes are in place, thank you!  I do have a followup question, though -- the new `lockDataIsIncomplete()` helper function doesn't test `... || last.who == ownerUnknown` the way the prior code did.

That is, the new tests in `lockDataIsIncomplete()` are:
```go
data.lock.Path == "" || data.lock.Owner == nil || data.lock.LockedAt.IsZero()
```

and the old tests were:
```go
last.lock.Path == "" || last.lock.Owner == nil || last.lock.LockedAt.IsZero() || last.who == ownerUnknown
```

Is that an intentional change, or just an oversight?

bk2204 replied:
This is intentional.  I found a test failure there which was hidden by the fact that we have only one lock.  The problem is that the function is used for both cases where we get an owner and cases where we don't, so if we check, then we fail in cases where we don't get one.

chrisd8088 replied:
Makes sense, just thought I'd check.

chrisd8088 reviewed (changes requested):
OK, round four is definitely getting down to a much smaller set of items!!  :-)

I've left three conversations from the last round open, with additional comments, and added just a few more in this latest review, almost all concentrated around a pair of functions in `ssh/protocol.go`, namely `SendMessage()` and `ReadStatusWithData()` and their callers.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,259 @@
+func (conn *PktlineConnection) End() error {
+	conn.Lock()
+	defer conn.Unlock()
+	err := conn.SendMessage("quit", nil, true)
```
I can definitely fix that.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,259 @@
+	return conn.pl.WriteFlush()
+}
+
+func (conn *PktlineConnection) ReadStatus() (int, []string, error) {
+	args := make([]string, 0, 100)
```
Yup, sounds good.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,259 @@
+				return 0, nil, nil, errors.NewProtocolError("no status seen", nil)
+			}
+			return status, args, nil, nil
+		} else if pktLen == 1 {
+			break
```
Done.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,386 @@
+	if err != nil {
+		return err
+	}
+	status, _, data, err := conn.ReadStatusWithData()
```
I think this adds a useful check here, especially if we turn an SSH server into a proxy for an HTTP server, as I suspect a bunch of people will do, so I'm going to verify the size argument.

chrisd8088 replied:
OK, cool!  I'm going to mark this resolved, and then separately add just one comment on the size-check implementation.

bk2204 commented on ssh/protocol.go:
```diff
@@ -0,0 +1,259 @@
+		if err != nil {
+			return 0, nil, nil, errors.NewProtocolError("error reading packet", err)
+		}
+		if pktLen == 0 {
+			if !seenStatus {
+				return 0, nil, nil, errors.NewProtocolError("no status seen", nil)
+			}
+			return status, args, nil, nil
```
No, I don't think that we should see a flush packet.  We should always see a delimiter packet first.

bk2204 commented on tq/ssh.go:
```diff
@@ -0,0 +1,386 @@
+	}
+	if status < 200 || status > 299 {
+		if data != nil {
+			io.Copy(ioutil.Discard, data)
```
Good point.  I've parsed that data up to 1024 bytes and then appended it to the error message.

chrisd8088 replied:
This may be overly cautious on my part ... since we're dealing with an error condition here, what would happen if the server returned a huge blob of data in the place of the `error-message`?  Would there be any sense in using `io.CopyN(buffer, data, 1024)` first, to try to get no more than 1024 bytes, and then if it returns `err` as `nil` (i.e., there's more to read), using  `io.Copy(ioutil.Discard, data)` to slurp the rest?

bk2204 replied:
Sure, that seems fine.  There are probably other fun ways the server can DoS us, but that doesn't mean we need to add to them.

bk2204 commented:
Okay, I think I should have gotten everything, but please tell me if I've missed something.

chrisd8088 commented on tq/ssh.go:
```diff
@@ -0,0 +1,405 @@
+	if bReq.Ref != nil {
+		args = append(args, fmt.Sprintf("refname=%s", bReq.Ref.Name))
+	}
+	status, args, lines, err := a.batchInternal(args, batchLines)
```
I believe the `args` return value is ignored after this, so this could be either:
```go
 	status, _, lines, err := a.batchInternal(args, batchLines)
```
 or `batchInternal()` could not even bother to return `args` at all, so:
```go
 	status, lines, err := a.batchInternal(args, batchLines)
```

If the latter case is preferred, then I'd further suggest that in `batchInternal()` we change to:
```go
 	status, _, lines, err := conn.ReadStatusWithLines()`
```

bk2204 replied:
I've just dropped the extra args in both places.

chrisd8088 replied:
Awesome, thank you.  I think there's one last tweak possible here, which would be to use `status, _, lines, err := conn.ReadStatusWithLines()` in `batchInternal()`, just to make it clear we're ignoring the `args` return value.

bk2204 replied:
Great, I've just fixed that.

chrisd8088 commented on tq/ssh.go:
```diff
@@ -0,0 +1,405 @@
+		return errors.NewRetriableError(fmt.Errorf("got status %d when fetching OID %s: %s", status, t.Oid, buffer.String()))
+	}
+
+	actualSize := int64(-1)
+	for _, arg := range args {
+		entries := strings.SplitN(arg, "=", 2)
+		if entries[0] == "size" {
+			val, err := strconv.ParseInt(entries[1], 10, 64)
+			if err != nil {
+				return fmt.Errorf("expected valid size, got %q", entries[1])
+			}
+			actualSize = val
+		}
+	}
```
With my "bad server" thinking cap on again, should we guard against multiple or invalid (e.g., negative) `size` argument packet lines here?

The proposed APi spec does [say](https://github.com/git-lfs/git-lfs/blob/05f7e0c8b8ee7378ebbb87b8fb5d655a871cf2b3/docs/proposals/ssh_adapter.md#requests-to-transfer-objects) that "unknown arguments should be ignored".  But since the `"size"` argument is known and expected, maybe something like the following would cover cases of extra, missing, and invalid ones?  (I'm unsure if a zero size is acceptable; I allowed it here.)

```suggestion
	var actualSize int64
	seenSize := false
	for _, arg := range args {
		if strings.HasPrefix(arg, "size=") {
			if seenSize {
				return errors.NewProtocolError("unexpected size argument", nil)
			}
			actualSize, err = strconv.ParseInt(arg[5:], 10, 64)
			if err != nil || actualSize < 0 {
				return errors.NewProtocolError(fmt.Sprintf("expected valid size, got %q", arg[5:]), err)
			}
			seenSize = true
		}
	}
	if !seenSize {
		return errors.NewProtocolError("no size argument seen", nil)
	}
```

bk2204 replied:
Sure.  I've squashed in your changes with a co-author credit.

chrisd8088 reviewed (changes requested):
Really down to the nitty-gritty at this point, thank you!

I think I've only spotted these remaining tweaks:
1. Changing back to `SendMessage(command string, args []string)` per my most recent comment in this https://github.com/git-lfs/git-lfs/pull/4446#discussion_r663448185 thread.
2. Maybe using `io.CopyN()` and then `io.Copy(ioutil.Discard, ...)` to slurp a potentially large `error-message`, per my most recent comment in this https://github.com/git-lfs/git-lfs/pull/4446#discussion_r671609337 thread.  If that seems overly paranoid, please ignore my comment.
3. In `Batch()` in `tq/ssh.go`, dropping the unused `args` from `batchInternal()`, either by just ignoring it, or not even returning it in the first place from that internal function.
4. In `doDownload()` in `tq/ssh.go`, I had some ideas for additional potential scanning for missing/extra/invalid `"size"` argument packet lines.

bk2204 commented:
I've fixed those four remaining things.

chrisd8088 reviewed (approved):
This is looking spectacular.  I've left one last conversation open (https://github.com/git-lfs/git-lfs/pull/4446#discussion_r672561036) where there's a lingering `args` which could be excised, but it's hardly urgent.

Thanks again for so much patience with all my questions, musings, and suggestions.

NeatNit commented:
Sorry if this is not the right place to ask, but I genuinely can't find anywhere else. As a user, how do I set up my client, repository, and server to use this? For example, I assume I need to define where on the server the large files will be stored.

Version 3.0.0 with this addition is scheduled for release this Friday, and I'd love to know how to use this before then.

slonopotamus commented:
As far as I understand, there isn't any existing git-lfs server that supports this feature.

estan commented:
FWIW, as soon as this was merged I filed https://gitlab.com/gitlab-org/gitlab/-/issues/336618 against GitLab, so if you happen to use GitLab to manage your repos, you could vote/chime in there.

ANogin commented:
> 
> 
> FWIW, as soon as this was merged I filed https://gitlab.com/gitlab-org/gitlab/-/issues/336618 against GitLab, so if you happen to use GitLab to manage your repos, you could vote/chime in there.

Is there a similar one for GitHub?

estan commented:
Not that I know of, but I haven't looked.

bk2204 commented:
As far as I know, there are no forges that support this feature.  https://github.com/bk2204/scutiger provides a `git-lfs-transfer` binary which can be used over SSH on a server using a normal Git repository as the backend.

dhy2000 commented:
Is there any example usage of LFS download/upload only via SSH?

And does GitHub, GitLab and other mainstream of Git service support it?

shamefulCake1 commented:
How do I set this up? Just add `git-lfs-transfer` into `$PATH`?

chrisd8088 commented:
> How do I set this up? Just add `git-lfs-transfer` into `$PATH`?

@shamefulCake1 -- I would recommend opening a new discussion if you're interested in this topic, as this PR has been merged and closed for a few years now.  Or you might find the answers you're looking for in an existing discussion like #5877.

In brief, the Git LFS client as of v3.0.0 doesn't need additional tools to perform the client-side portions of the pure SSH-based Git LFS transfer protocol.  However, not all Git LFS hosting services offer support for the server-side portions of the protocol yet, so the answer to your question depends a lot on what service you are using.  GitLab supports the protocol now, and I believe Gitea has added support as well, but GitHub has not.

If you are setting up your own Git LFS hosting service, you could use the `git-lfs-transfer` binary from the `bk2204/scutiger` [project](https://github.com/bk2204/scutiger/tree/614a4f01cfbe6aafa355ca92ed6d1581b5cfc088/scutiger-lfs#git-lfs-transfer) along with the rest of your Git and Git LFS services, or you could write your own implementation, as appropriate, so long as it conforms to the [proposal](https://github.com/git-lfs/git-lfs/blob/a577e336ebdccfd312b6006c880f010b5d3fe796/docs/proposals/ssh_adapter.md) for the protocol.

shamefulCake1 commented:
Thank you, it seems that https://github.com/git-lfs/git-lfs/discussions/5877 is indeed what I need. 

I recognise that this PR is several years old, but it's still the first one to come up on Google and DDG.

>If you are setting up your own Git LFS hosting service, 

I'm not doing anything even remotely so complicated. I'm just trying to clone my a git repo from my workstation to my my laptop over ssh. Git is a distributed VCS, and it doesn't need a server.
